### PR TITLE
Task API app benchmarks

### PR DESCRIPTION
**Pull request by mfranciszkiewicz**
Replaces `env` benchmarks with `app` benchmarks in WTCT.
- store app benchmarks in a separate db table
- do not compute a benchmark if one is already running
- clear app benchmarks on configuration change

### CODE REVIEW

File: golem/task/benchmarkmanager.py [hunk 1]
```diff
@@ -129,3 +138,75 @@ def run_default_benchmark(callback, errback):
...
+                comp_type=ComputationType.BENCHMARK,
+                comp_id=self._computing)
+
+        try:
+            self._computing = prereq_hash
+            score = await self._run_benchmark(env_id, env_prereq_dict)
+        finally:
+            self._computing = None
+
+        benchmark = AppBenchmark(hash=prereq_hash, score=score)
```
Comment (maaktweluit) [inline golem/task/benchmarkmanager.py:186]: just to check: will this run when the benchmark raises an error?

File: golem/task/benchmarkmanager.py [hunk 2]
```diff
@@ -129,3 +138,75 @@ def run_default_benchmark(callback, errback):
...
+    async def _run_benchmark(
+            self,
+            env_id: EnvId,
+            env_prereq_dict: Dict[str, Any]
+    ) -> float:
+        env = self._env_manager.environment(env_id)
+        prereq_hash = hash_prereq_dict(env_prereq_dict)
+
+        shared_dir = self._root_path / prereq_hash
+        shared_dir.mkdir(parents=True, exist_ok=True)
```
Comment (maaktweluit) [inline golem/task/benchmarkmanager.py:203]: should we clean up these folders after the benchmark is completed?

File: golem/task/taskserver.py [hunk 3]
```diff
@@ -404,15 +411,22 @@ def _request_task(self, theader: dt_tasks.TaskHeader) -> Deferred:
...
-                logger.debug("Not requesting task, benchmark is in progress.")
-                return None
+                try:
+                    future = self.app_benchmark_manager.get_benchmark_score(
+                        theader.environment,
+                        theader.environment_prerequisites)
+                    app_benchmark = yield Deferred.fromFuture(
+                        asyncio.ensure_future(future))
+                except ComputationInProgress as e:
+                    logger.debug(f"Not requesting task={theader.task_id}: {e}")
```
Comment (maaktweluit) [inline golem/task/taskserver.py]: better to use the native formatter of `logger`

File: tests/golem/task/test_appbenchmarkmanager.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
+# pylint: disable=unused-argument
+# pylint: disable=redefined-outer-name
+# ^^ Pytest fixtures in the same file require the same name
```
Comment (maaktweluit) [inline tests/golem/task/test_appbenchmarkmanager.py:3]: maybe we can move these to the `lint.sh`

https://github.com/golemfactory/golem/blob/develop/lint.sh#L103

File: tests/golem/task/test_appbenchmarkmanager.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+        assert self.app_benchmark_manager._run_benchmark.called
+        assert score == 10.
+
+    @pytest.mark.asyncio
+    async def test_get_benchmark_score_exception(self):
+        self.app_benchmark_manager._run_benchmark = AsyncMock(return_value=10.)
+        self.app_benchmark_manager._computing = True
+
+        try:
```
Comment (maaktweluit) [inline tests/golem/task/test_appbenchmarkmanager.py]: you can use `with pytest.raises(...):` here:

http://doc.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions

**Review by maaktweluit — CHANGES_REQUESTED** [review]
LGTM!

Some small comments, will approve on reply or fix

File: golem/task/benchmarkmanager.py (hunk 1 above)
Comment (mfranciszkiewicz) [reply golem/task/benchmarkmanager.py:186]: It won't, exceptions aren't caught in the `try ... finally` block

File: tests/golem/task/test_appbenchmarkmanager.py (hunk 4 above)
Comment (mfranciszkiewicz) [reply tests/golem/task/test_appbenchmarkmanager.py:3]: I'd rather not touch `lint.sh` here

**Review by maaktweluit — APPROVED** [review]

File: golem/task/taskserver.py [hunk 6]
```diff
@@ -404,15 +411,25 @@ def _request_task(self, theader: dt_tasks.TaskHeader) -> Deferred:
...
-                env_mgr = self.task_keeper.new_env_manager
-                benchmark_result = yield env_mgr.get_benchmark_result(env_id)
-            if benchmark_result is None:
-                logger.debug("Not requesting task, benchmark is in progress.")
-                return None
+                try:
+                    future = self.app_benchmark_manager.get_benchmark_score(
+                        theader.environment,
+                        theader.environment_prerequisites)
+                    app_benchmark = yield Deferred.fromFuture(
```
Comment (Wiezzel) [inline golem/task/taskserver.py]: Plz, use `deferred_from_future` from `golem.core.deferred`. It also calls `ensure_future`.

File: golem/task/taskserver.py [hunk 7]
```diff
@@ -404,15 +411,25 @@ def _request_task(self, theader: dt_tasks.TaskHeader) -> Deferred:
...
+                        asyncio.ensure_future(future))
+                except ComputationInProgress as error:
```
Comment (Wiezzel) [inline golem/task/taskserver.py:417]: Other errors could also occur, like something gone wrong in benchmark computation. Don't we need any handling for these?

File: golem/model.py [hunk 8]
```diff
@@ -574,6 +574,16 @@ def update_or_create(env_id: str, performance: float, cpu_usage: int):
             ).save()
 
 
+class AppBenchmark(BaseModel):
+
+    hash = CharField(null=False, index=True, unique=True)
+    score = FloatField()
+    cpu_usage = IntegerField(default=0)  # total CPU usage in nanoseconds
```
Comment (Wiezzel) [inline golem/model.py]: Wouldn't default 0 cause any troubles (vide #4811)?

**Pull request merged**